File: app/src/main/java/com/wolcano/musicplayer/music/ui/adapter/GenreAdapter.java
package com.wolcano.musicplayer.music.ui.adapter;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.wolcano.musicplayer.music.R;
import com.wolcano.musicplayer.music.databinding.ItemGenreBinding;
import com.wolcano.musicplayer.music.mvp.models.Genre;
import com.wolcano.musicplayer.music.utils.Utils;
import java.util.List;


public class GenreAdapter extends RecyclerView.Adapter<GenreAdapter.ViewHolder> {

    private List<Genre> arraylist;
    private Context context;

    public GenreAdapter(Context context, List<Genre> arraylist){
        this.context = context;
        this.arraylist = arraylist;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemGenreBinding binding = DataBindingUtil.inflate(LayoutInflater.from(parent.getContext()), R.layout.item_genre, parent, false);
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {

        holder.binding.setGenre(arraylist.get(position));
        holder.binding.executePendingBindings();
        Genre genre = holder.binding.getGenre();
        holder.binding.line1.setText(genre.getName());
        holder.binding.line2.setText(Utils.createStr(context, R.plurals.Nsongs, genre.getSongCount()));

    }

    @Override
    public int getItemCount() {
        return (null != arraylist ? arraylist.size() : 0);
    }


    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        private ItemGenreBinding binding;

        public ViewHolder(ItemGenreBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
            this.binding.getRoot().setOnClickListener(this::onClick);
        }

        @Override
        public void onClick(View v) {

           Genre genre = arraylist.get(getAdapterPosition());
           long genreId = genre.getId();
           String genreName = genre.getName();
           Utils.navigateToGenre(context, genreId, genreName);

        }
    }
}

File: app/src/main/java/com/wolcano/musicplayer/music/ui/adapter/GenreAdapter.kt
package com.wolcano.musicplayer.music.ui.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import com.wolcano.musicplayer.music.R
import com.wolcano.musicplayer.music.databinding.ItemGenreBinding
import com.wolcano.musicplayer.music.data.model.Genre
import com.wolcano.musicplayer.music.utils.Utils.createStr
import com.wolcano.musicplayer.music.utils.Utils.navigateToGenre

class GenreAdapter(private val context: Context, private val genreList: List<Genre>) :
    RecyclerView.Adapter<GenreAdapter.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): GenreAdapter.ViewHolder {
        val binding: ItemGenreBinding = DataBindingUtil.inflate(
            LayoutInflater.from(parent.context),
            R.layout.item_genre,
            parent,
            false
        )
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: GenreAdapter.ViewHolder, position: Int) {
        holder.binding.genre = genreList[position]
        holder.binding.executePendingBindings()

        val genre: Genre? = holder.binding.genre

        holder.binding.line1.text = genre?.name
        holder.binding.line2.text = createStr(context, R.plurals.Nsongs, genre!!.songCount)
    }

    override fun getItemCount(): Int {
        return genreList.size
    }

    inner class ViewHolder(val binding: ItemGenreBinding) :
        RecyclerView.ViewHolder(binding.root),
        View.OnClickListener {
        override fun onClick(v: View) {
            val genre: Genre = genreList[absoluteAdapterPosition]
            val genreId = genre.id
            val genreName = genre.name
            navigateToGenre(context, genreId, genreName)
        }

        init {
            this.binding.root.setOnClickListener(::onClick)
        }
    }

}